File: android/app/src/main/java/co/slidepress/app/NativeWidgetPlugin.kt
package co.slidepress.app

import android.util.Log
import co.slidepress.app.widget.WidgetSnapshotStore
import co.slidepress.app.widget.WidgetUpdateHelper
import com.getcapacitor.Plugin
import com.getcapacitor.PluginCall
import com.getcapacitor.PluginMethod
import com.getcapacitor.annotation.CapacitorPlugin

@CapacitorPlugin(name = "NativeWidget")
class NativeWidgetPlugin : Plugin() {
    @PluginMethod
    fun setSnapshot(call: PluginCall) {
        val snapshot = call.getString("snapshot")

        if (snapshot.isNullOrBlank()) {
            call.reject("snapshot is required")
            return
        }

        val appContext = context.applicationContext
        WidgetSnapshotStore.saveSnapshotJson(appContext, snapshot)
        Log.d(TAG, "Saved widget snapshot (${snapshot.length} chars)")
        WidgetUpdateHelper.refreshAllWidgets(appContext)
        call.resolve()
    }

    @PluginMethod
    fun clearSnapshot(call: PluginCall) {
        val appContext = context.applicationContext
        WidgetSnapshotStore.clearSnapshot(appContext)
        Log.d(TAG, "Cleared widget snapshot")
        WidgetUpdateHelper.refreshAllWidgets(appContext)
        call.resolve()
    }

    companion object {
        private const val TAG = "SlidePressWidget"
    }
}

File: ios/App/App/NativeWidgetPlugin.swift
import Capacitor
import WidgetKit

@objc(NativeWidgetPlugin)
public class NativeWidgetPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "NativeWidgetPlugin"
    public let jsName = "NativeWidget"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "setSnapshot", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "clearSnapshot", returnType: CAPPluginReturnPromise),
    ]

    @objc func setSnapshot(_ call: CAPPluginCall) {
        guard let snapshot = call.getString("snapshot") else {
            call.reject("snapshot is required")
            return
        }

        WidgetSnapshotStore.saveSnapshotJson(snapshot)

        if #available(iOS 14.0, *) {
            WidgetCenter.shared.reloadAllTimelines()
        }

        call.resolve()
    }

    @objc func clearSnapshot(_ call: CAPPluginCall) {
        WidgetSnapshotStore.clearSnapshot()

        if #available(iOS 14.0, *) {
            WidgetCenter.shared.reloadAllTimelines()
        }

        call.resolve()
    }
}
